helix-term/src/handlers/auto_save.rs

```diff
@@ -4,36 +4,68 @@ use anyhow::Ok;
 use arc_swap::access::Access;
 
 use helix_event::{register_hook, send_blocking};
-use helix_view::{events::DocumentDidChange, handlers::Handlers, Editor};
+use helix_view::{
+    document::Mode,
+    events::DocumentDidChange,
+    handlers::{AutoSaveEvent, Handlers},
+    Editor,
+};
 use tokio::time::Instant;
 
 use crate::{
     commands, compositor,
+    events::OnModeSwitch,
     job::{self, Jobs},
 };
 
 #[derive(Debug)]
-pub(super) struct AutoSaveHandler;
+pub(super) struct AutoSaveHandler {
+    save_pending: bool,
+}
 
 impl AutoSaveHandler {
     pub fn new() -> AutoSaveHandler {
-        AutoSaveHandler
+        AutoSaveHandler {
+            save_pending: false,
+        }
     }
 }
 
 impl helix_event::AsyncHook for AutoSaveHandler {
-    type Event = u64;
+    type Event = AutoSaveEvent;
 
     fn handle_event(
         &mut self,
-        timeout: Self::Event,
-        _: Option<tokio::time::Instant>,
+        event: Self::Event,
+        existing_debounce: Option<tokio::time::Instant>,
     ) -> Option<Instant> {
-        Some(Instant::now() + Duration::from_millis(timeout))
+        match event {
+            Self::Event::DocumentChanged { debounce } => {
+                self.save_pending = true;
```

Comment:
as far as I can see we never reset `save_pending` back to false. I guess that would need to happen when leaving insertmodel? This also sets this to true in normal mode would also cause excess/incorrect saves. 

I actually had something else in mind:

* make `save_pending` an `Arc<AtomicBool>`, only inside finish_deboucne/dispatch_blocking if the mode is insert mode store true in there (needs to be an Arc/AtomicBool so we can move it inside the callback)
* Another clone of the arc is moved into the mode switch callback. When going from insert to normal mode that callback will check the value of the shared bool. If it's true it will instantly save (not send an even, save right in the mode switch handler) and then store false in the callback

I think doing it as an event that triggers a new save could work too. But I think the Arc/AtomicBool that only gets set to true if the calblack actually failed (so only we we are actually in normal mode when trying to save) is definitly necessary

Reply:
Ah yep I forgot to reset it. But I like the sound of the atomic bool more!

Reply:
I switched to the atomic bool so we can set it in the dispatch callback but I didn't also do the part about passing the bool to the mode switch hook. I'd like to keep the behavior where we delay even when switching modes rather than saving immediately on mode switch.

Maybe auto-save on mode switch should be a separate auto-save option though?

Reply:
hmm so I wasn't necessarily thinking of always saving immediately on mode switch but rather:

* the save should happen in insert-mode, the timeout already expired we just can't save right now due to internal limitations so we set a flag that an autosave is "pending"
* as soon as we exiting insertmode we execute this "pending" write right away. We execute it right aways since it really should have executed earlier in insertmode, we just couldn't execute it due to internal limitations so we execute it first thing we get

(the pending flag would be the `Arc<AtomicBool>`)

this is kind of orthogonal to saving on mode switch/exiting insertmode in general. The main difference is that it's a bit more robust by not passing trough the timeout mechanism. 

On the other hand I am not sure if savin in the mode switch hook is even possible. I think we only create the history entry after the mode switch hook executes so going trough the event handler is probably necessary/the right thing to do.

Reply:
Oh I see now, I misunderstood. I was also curious about the ordering and I believe you're correct that we need to go through the event handler so that we save to history before attempting to write.